File: app/src/main/java/com/tevinjeffrey/rutgersct/data/notifications/UCTFirebaseMessagingService.java
/**
 * Copyright 2015 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tevinjeffrey.rutgersct.data.notifications;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.support.v4.app.NotificationCompat;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import com.crashlytics.android.answers.Answers;
import com.crashlytics.android.answers.CustomEvent;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.google.gson.Gson;
import com.tevinjeffrey.rutgersct.R;
import com.tevinjeffrey.rutgersct.data.model.Course;
import com.tevinjeffrey.rutgersct.data.model.Section;
import com.tevinjeffrey.rutgersct.data.model.Subject;
import com.tevinjeffrey.rutgersct.data.model.UCTNotification;
import com.tevinjeffrey.rutgersct.data.model.University;
import com.tevinjeffrey.rutgersct.utils.PreferenceUtils;
import dagger.android.AndroidInjection;
import dagger.android.AndroidInjector;
import java.io.IOException;
import java.math.BigInteger;
import javax.inject.Inject;
import timber.log.Timber;

public class UCTFirebaseMessagingService extends FirebaseMessagingService {

  private static final String SECTION_NOTIFICATION_GROUP = "SECTION_NOTIFICATION_GROUP";
  private static final String TAG = "FirebaseMessagingSer";
  @Inject
  Gson gson;
  @Inject
  PreferenceUtils mPreferenceUtils;

  @Override public void onCreate() {
    AndroidInjection.inject(this);
    super.onCreate();
  }

  public void onMessageReceived(RemoteMessage message) {
    String data = message.getData().get("message");
    Timber.d(new StringBuilder().append("From: ").append(message.getFrom()).toString());
    Timber.d(new StringBuilder().append("Message: ").append(data).toString());

    if (data == null) {
      handleGenericNotification(message);
      return;
    }

    sendNotification(data);
  }

  private Uri getSound() {
    if (mPreferenceUtils.getCanPlaySound()) {
      return RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION);
    } else {
      return null;
    }
  }

  private void handleGenericNotification(RemoteMessage message) {
  }

  private void sendNotification(String message) {
    UCTNotification uctNotification = null;
    try {
      uctNotification = gson.getAdapter(UCTNotification.class).fromJson(message);
    } catch (IOException e) {
      e.printStackTrace();
    }

    University university = uctNotification.university;
    Subject subject = university.subjects.get(0);
    Course course = subject.courses.get(0);
    Section section = course.sections.get(0);

    NotificationManager notificationManager =
        (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
    NotificationCompat.Builder mBuilder;

    String title;
    String body;
    int color;
    if (uctNotification.status.equals("Open")) {
      title = "A section has opened!";
      body = "Section " + section.number + " of " + course.name + " has opened!";
      color = R.color.green;
    } else {
      title = "A section has closed!";
      body = "Section " + section.number + " of " + course.name + " has closed!";
      color = R.color.red;
    }

    mBuilder =
        new NotificationCompat.Builder(this)
            .setStyle(new NotificationCompat.BigTextStyle()
                .bigText(body)
                .setBigContentTitle(course.name))
            .setSmallIcon(R.drawable.ic_notification)
            .setWhen(System.currentTimeMillis())
            .setPriority(NotificationCompat.PRIORITY_MAX)
            .setGroup(SECTION_NOTIFICATION_GROUP)
            .setColor(ContextCompat.getColor(this, color))
            .setAutoCancel(true)
            .setSound(getSound())
            .setContentTitle(title)
            .setContentText(body);

    //Intent to start web browser
    Intent openInBrowser = new Intent(Intent.ACTION_VIEW);
    openInBrowser.setData(Uri.parse("https://sims.rutgers.edu/webreg/"));
    PendingIntent pOpenInBrowser = PendingIntent.getActivity(this, 0, openInBrowser, 0);
    mBuilder.addAction(R.drawable.ic_open_in_browser, "Webreg", pOpenInBrowser);
    mBuilder.setContentIntent(pOpenInBrowser);

    notificationManager.notify(
        new BigInteger(String.valueOf(System.currentTimeMillis())).intValue(),
        mBuilder.build()
    );

    Answers.getInstance()
        .logCustom(new CustomEvent("receive_notification")
            .putCustomAttribute("status", section.status)
            .putCustomAttribute("topic", section.topic_name));

    Log.d("FCM message", message);
  }

  @dagger.Subcomponent
  public interface Subcomponent extends AndroidInjector<UCTFirebaseMessagingService> {
    @dagger.Subcomponent.Builder
    abstract class Builder extends AndroidInjector.Builder<UCTFirebaseMessagingService> {
      @Override public abstract Subcomponent build();
    }
  }
}

File: app/src/main/java/com/tevinjeffrey/rutgersct/data/notifications/UCTFirebaseMessagingService.kt
package com.tevinjeffrey.rutgersct.data.notifications

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationChannelGroup
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.media.RingtoneManager
import android.net.Uri
import android.os.Build
import android.support.v4.app.NotificationCompat
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.squareup.moshi.Moshi
import com.tevinjeffrey.rutgersct.R
import com.tevinjeffrey.rutgersct.data.UCTApi
import com.tevinjeffrey.rutgersct.ui.MainActivity
import com.tevinjeffrey.rutgersct.utils.PreferenceUtils
import dagger.android.AndroidInjection
import io.reactivex.android.schedulers.AndroidSchedulers
import timber.log.Timber
import java.math.BigInteger
import javax.inject.Inject

class UCTFirebaseMessagingService : FirebaseMessagingService() {
  @Inject lateinit var moshi: Moshi
  @Inject lateinit var uctApi: UCTApi
  @Inject lateinit var mPreferenceUtils: PreferenceUtils

  private lateinit var notificationManager: NotificationManager

  private val sound: Uri?
    get() = if (mPreferenceUtils.canPlaySound) {
      RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
    } else {
      null
    }

  override fun onCreate() {
    AndroidInjection.inject(this)
    super.onCreate()
    notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
      val group = createSectionGroup()
      createOpenChannel(group)
      createClosedChannel(group)
      createGenericChannel()
    }
  }

  @TargetApi(Build.VERSION_CODES.O)
  private fun createSectionGroup(): String {
    // The id of the group.
    val group = getString(R.string.section_group);
    // The user-visible name of the group.
    val groupName = getString(R.string.sections)
    notificationManager.createNotificationChannelGroup(NotificationChannelGroup(group, groupName))
    return group
  }

  @TargetApi(Build.VERSION_CODES.O)
  private fun createGenericChannel() {
    val id = getString(R.string.channel_generic)
    val name = getString(R.string.other)
    val importance = NotificationManager.IMPORTANCE_DEFAULT
    val channel = NotificationChannel(id, name, importance)
    notificationManager.createNotificationChannel(channel)
  }

  @TargetApi(Build.VERSION_CODES.O)
  private fun createClosedChannel(group: String) {
    val id = getString(R.string.channel_section_closed)
    val name = getString(R.string.channel_closed_name)
    val description = getString(R.string.channel_closed_description)
    val importance = NotificationManager.IMPORTANCE_HIGH
    val channel = NotificationChannel(id, name, importance)
    channel.description = description
    channel.enableLights(true)
    channel.enableVibration(true)
    channel.group = group
    notificationManager.createNotificationChannel(channel)
  }

  @TargetApi(Build.VERSION_CODES.O)
  private fun createOpenChannel(group: String) {
    val id = getString(R.string.channel_section_open)
    val name = getString(R.string.channel_open_name)
    val description = getString(R.string.channel_open_description)
    val importance = NotificationManager.IMPORTANCE_HIGH
    val channel = NotificationChannel(id, name, importance)
    channel.description = description
    channel.enableLights(true)
    channel.enableVibration(true)
    channel.group = group
    notificationManager.createNotificationChannel(channel)
  }

  override fun onMessageReceived(message: RemoteMessage) {
    if (message.data["notificationId"] == null) {
      handleGenericNotification(message)
      return
    }

    val title: String = message.data["title"].orEmpty()
    val body: String = message.data["body"].orEmpty()
    val topicName: String = message.data["topicName"].orEmpty()
    val topicId: String = message.data["topicId"].orEmpty()
    val notificationId: String = message.data["notificationId"].orEmpty()
    val status: String = message.data["status"].orEmpty()
    val color: String = message.data["color"].orEmpty()
    val registrationUrl: String = message.data["registrationUrl"].orEmpty()
    sendNotification(
        title = title,
        body =  body,
        color = color,
        topicName = topicName,
        topicId = topicId,
        status = status,
        notificationId = notificationId,
        registrationUrl = registrationUrl
    )
  }

  private fun handleGenericNotification(message: RemoteMessage) {
    val notificationBuilder = NotificationCompat.Builder(this, getString(R.string.channel_generic))
        .setSmallIcon(R.drawable.ic_notification)
        .setWhen(System.currentTimeMillis())
        .setContentTitle(message.notification?.title)
        .setContentText(message.notification?.body)

    val intent = openAppIntent()
    notificationBuilder.setContentIntent(intent)

    val notification = notificationBuilder.build()
    notification.flags = notification.flags or Notification.FLAG_AUTO_CANCEL

    message.notification
    notificationManager.notify(0, notification)
  }

  private fun openAppIntent(): PendingIntent {
    val notificationIntent = Intent(this.applicationContext, MainActivity::class.java)
    notificationIntent.flags = Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP
    return PendingIntent.getActivity(applicationContext, 0, notificationIntent, 0)
  }

  private fun sendNotification(title: String,
                               body: String,
                               status: String,
                               color: String,
                               topicId: String,
                               topicName: String,
                               notificationId: String,
                               registrationUrl: String) {

    val channel = if (status == "Open") {
      getString(R.string.channel_section_open)
    } else {
      getString(R.string.channel_section_closed)
    }

    val notificationBuilder = NotificationCompat.Builder(this, channel)
        .setSmallIcon(R.drawable.ic_notification)
        .setWhen(System.currentTimeMillis())
        .setPriority(NotificationCompat.PRIORITY_MAX)
        .setGroup(getString(R.string.section_notification_group))
        .setColor(Color.parseColor(color))
        .setAutoCancel(true)
        .setSound(sound)
        .setContentTitle(title)
        .setContentText(body)

    //Intent to start web browser
    val actionIntent = PendingIntent.getActivity(this, 0, Intent(Intent.ACTION_VIEW).apply {
      data = Uri.parse(registrationUrl)
    }, 0)

    notificationBuilder
        .addAction(R.drawable.ic_open_in_browser, getString(R.string.register), actionIntent)
        .setContentIntent(openAppIntent())

    notificationManager.notify(
        BigInteger(System.currentTimeMillis().toString()).toInt(),
        notificationBuilder.build()
    )

    uctApi.acknowledgeNotification(
        topicId,
        topicName,
        notificationId).observeOn(AndroidSchedulers.mainThread())
        .subscribe({
          Timber.i("Acknowledged notification: $notificationId for topic: $topicName")
        }, { Timber.e(it) })
  }
}
